pyvrp/accept/MovingAverageThreshold.py

```diff
@@ -0,0 +1,151 @@
+        the threshold value. Must be positive.
+    max_runtime
+        Maximum runtime in seconds. As the search approaches this time limit,
+        :math:`w \to 0`. Must be non-negative. Default is ``None``, meaning
+        that :math:`w` stays equal to :math:`w_0`.
+    max_iterations
+        Maximum number of iterations. As the search approaches this limit,
+        :math:`w \to 0`. Must be non-negative. Default is ``None``, meaning
+        that :math:`w` stays equal to :math:`w_0`.
+
+    References
+    ----------
+    .. [1] Máximo, V.R. and M.C.V. Nascimento. 2021. A hybrid adaptive iterated
+           local search with diversification control to the capacitated vehicle
+           routing problem, *European Journal of Operational Research* 294 (3):
+           1108 - 1119.
+    """
+
+    def __init__(
+        self,
+        weight: float,
+        history_length: int,
+        max_runtime: float | None = None,
+        max_iterations: int | None = None,
+    ):
+        if weight < 0:
+            raise ValueError("weight must be non-negative.")
+
+        if history_length <= 0:
+            raise ValueError("history_length must be positive.")
+
+        if max_runtime is not None and max_runtime < 0:
+            raise ValueError("max_runtime must be non-negative.")
+
+        if max_iterations is not None and max_iterations < 0:
+            raise ValueError("max_iterations must be non-negative.")
+
+        self._weight = weight
+        self._history_length = history_length
+        self._max_runtime = max_runtime
+        self._max_iterations = max_iterations
+
+        self._history = np.zeros(history_length)
+        self._start_time = perf_counter()
+        self._iters = 0
+
+    @property
+    def weight(self) -> float:
+        return self._weight
+
+    @property
+    def history_length(self) -> int:
+        return self._history_length
+
+    @property
+    def max_runtime(self) -> float | None:
+        return self._max_runtime
+
+    @property
+    def max_iterations(self) -> float | None:
+        return self._max_iterations
+
+    @property
+    def _runtime_budget(self) -> float:
+        """
+        Returns the remaining runtime budget as percentage of the maximum
+        runtime.
+        """
+        if self.max_runtime is None:
+            return 1
+
+        if self.max_runtime == 0:
+            return 0
+
+        runtime = perf_counter() - self._start_time
+        return 1 - runtime / self.max_runtime
+
+    @property
+    def _iteration_budget(self) -> float:
+        """
+        Returns the remaining iteration budget as percentage of the maximum
+        number of iterations.
+        """
+        if self.max_iterations is None:
+            return 1
+
+        if self.max_iterations == 0:
+            return 0
+
+        return 1 - self._iters / self.max_iterations
+
+    def __call__(self, best: float, current: float, candidate: float) -> bool:
```

Comment:
This gets called quite a bit, right? For performance, it's then probably best to go through the variables directly, where possible, rather than their property members. So `self._history_length` rather than `self.history_length`, and `self._weight`.